### ISSUE TITLE
Remote PowerShell fails with "The service cannot accept control messages at this time.", then "The service has not been started."

### ISSUE THREAD

**Issue by qJake** [Microsoft/azure-pipelines-tasks#8262]
## Environment

**Server:** Team Foundation Server On-Premise, Version 16.131.27701.1

**Agent Info:**

Property | Value
----------|---------
(Type) | On-premise / Local Agent
Agent.Version | 2.131.0
Agent.OSVersion | 6.1.7601
AzurePS | 5.7.0
PowerShell | 5.1.14409.1012

## Issue Description

Our release definition includes, among other things, these steps:

1. Azure File Copy - Copies deployment files to a temp folder on the target machine
2. Azure File Copy - Copies a PowerShell script to the target machine
3. PowerShell on Target Machines - Invokes the copied PowerShell script to move and install the files, and perform some IIS tasks (App pool start/stop, etc)

This is being deployed to an Azure VM (Standard B1ms (1 vcpus, 2 GB memory), Windows Server 2016 Datacenter, PowerShell Version 5.1.14393.2312)

When running the task **PowerShell on Target Machines**, I get this series of events after a target machine fresh boot:

1. First deployment: Success
2. Second deployment: `The service cannot accept control messages at this time. (Exception from HRESULT: 0x80070425)`
3. Third deployment: `The service has not been started. (Exception from HRESULT: 0x80070426)`

It will keep repeating the third error each time I retry the deployment, until I reboot the target machine.

Any ideas on how to fix this?

### Log from Second Deployment

```
2018-09-05T21:25:20.9776978Z ==============================================================================
2018-09-05T21:25:20.9776978Z Task         : PowerShell on Target Machines
2018-09-05T21:25:20.9776978Z Description  : Execute PowerShell scripts on remote machine(s)
2018-09-05T21:25:20.9776978Z Version      : 2.0.3
2018-09-05T21:25:20.9776978Z Author       : Microsoft Corporation
2018-09-05T21:25:20.9776978Z Help         : [More Information](https://go.microsoft.com/fwlink/?linkid=627414)
2018-09-05T21:25:20.9776978Z ==============================================================================
2018-09-05T21:25:23.4113134Z Deployment started for machine: '<<<TARGET AZURE MACHINE OMITTED>>>:5986'
2018-09-05T21:25:32.2565701Z Deployment logs for Deployment operation on <<<TARGET AZURE MACHINE OMITTED>>>:5986 
2018-09-05T21:25:32.2565701Z 
2018-09-05T21:25:32.2565701Z 	 Deployment started on target machine...
2018-09-05T21:25:32.2565701Z 
2018-09-05T21:25:32.2565701Z System.AggregateException: Failed to execute the powershell script. Consult the logs below for details of the error.
2018-09-05T21:25:32.2565701Z The service cannot accept control messages at this time. (Exception from HRESULT: 0x80070425) ---> System.Management.Automation.RuntimeException: The service cannot accept control messages at this time. (Exception from HRESULT: 0x80070425) ---> System.Runtime.InteropServices.COMException: The service cannot accept control messages at this time. (Exception from HRESULT: 0x80070425)
2018-09-05T21:25:32.2565701Z    at Microsoft.IIs.PowerShell.Framework.Interop.IAppHostMethodInstance.Execute()
2018-09-05T21:25:32.2565701Z    at Microsoft.IIs.PowerShell.Framework.CodeMethod.Invoke(Object[] arguments)
2018-09-05T21:25:32.2565701Z    at System.Dynamic.UpdateDelegates.UpdateAndExecute1[T0,TRet](CallSite site, T0 arg0)
2018-09-05T21:25:32.2565701Z    at System.Management.Automation.Interpreter.DynamicInstruction`2.Run(InterpretedFrame frame)
2018-09-05T21:25:32.2565701Z    at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)
2018-09-05T21:25:32.2565701Z    --- End of inner exception stack trace ---
2018-09-05T21:25:32.2565701Z    at System.Management.Automation.Runspaces.PipelineBase.Invoke(IEnumerable input)
2018-09-05T21:25:32.2565701Z    at System.Management.Automation.PowerShell.Worker.ConstructPipelineAndDoWork(Runspace rs, Boolean performSyncInvoke)
2018-09-05T21:25:32.2565701Z    at System.Management.Automation.PowerShell.Worker.CreateRunspaceIfNeededAndDoWork(Runspace rsToUse, Boolean isSync)
2018-09-05T21:25:32.2565701Z    at System.Management.Automation.PowerShell.CoreInvokeHelper[TInput,TOutput](PSDataCollection`1 input, PSDataCollection`1 output, PSInvocationSettings settings)
2018-09-05T21:25:32.2565701Z    at System.Management.Automation.PowerShell.CoreInvoke[TInput,TOutput](PSDataCollection`1 input, PSDataCollection`1 output, PSInvocationSettings settings)
2018-09-05T21:25:32.2565701Z    at System.Management.Automation.PowerShell.Invoke(IEnumerable input, PSInvocationSettings settings)
2018-09-05T21:25:32.2565701Z    at Microsoft.VisualStudio.Services.DevTestLabs.RemoteDeployer.Deployment.PowershellExecutor.<>c__DisplayClass28_0.<Invoke>b__0()
2018-09-05T21:25:32.2565701Z    at Microsoft.VisualStudio.Services.DevTestLabs.RemoteDeployer.Utilities.RetryExecutor.<>c__DisplayClass19_0`1.<Execute>b__0()
2018-09-05T21:25:32.2565701Z    at Microsoft.VisualStudio.Services.DevTestLabs.RemoteDeployer.Utilities.RetryExecutor.Execute(Action action)
2018-09-05T21:25:32.2565701Z    at Microsoft.VisualStudio.Services.DevTestLabs.RemoteDeployer.Utilities.RetryExecutor.Execute[T](Func`1 action)
2018-09-05T21:25:32.2565701Z    at Microsoft.VisualStudio.Services.DevTestLabs.RemoteDeployer.Deployment.PowershellExecutor.Invoke(String errorContextMessage, Boolean writeResultToLog, Boolean isCancellable, ActionPreference errorActionPreference)
2018-09-05T21:25:32.2565701Z    --- End of inner exception stack trace ---
2018-09-05T21:25:32.2565701Z    at Microsoft.VisualStudio.Services.DevTestLabs.RemoteDeployer.Deployment.PowershellExecutor.Invoke(String errorContextMessage, Boolean writeResultToLog, Boolean isCancellable, ActionPreference errorActionPreference)
2018-09-05T21:25:32.2565701Z    at Microsoft.VisualStudio.Services.DevTestLabs.RemoteDeployer.Deployment.ScriptExecutor.ExecuteScript(IPowerShell powerShell, String script, String arguments, Boolean isScriptDotSourced, Boolean useHttp)
2018-09-05T21:25:32.2565701Z    at Microsoft.VisualStudio.Services.DevTestLabs.RemoteDeployer.Deployment.DeploymentService.RunPowerShellScripts(DeploymentMachineSpecification deploymentMachine, ScriptSpecification scriptSpecification, ScriptSpecification initializationScriptSpecification, String applicationPath, IPowerShell powerShellSession)
2018-09-05T21:25:32.2565701Z ---> (Inner Exception #0) System.Management.Automation.RuntimeException: The service cannot accept control messages at this time. (Exception from HRESULT: 0x80070425) ---> System.Runtime.InteropServices.COMException: The service cannot accept control messages at this time. (Exception from HRESULT: 0x80070425)
2018-09-05T21:25:32.2565701Z    at Microsoft.IIs.PowerShell.Framework.Interop.IAppHostMethodInstance.Execute()
2018-09-05T21:25:32.2565701Z    at Microsoft.IIs.PowerShell.Framework.CodeMethod.Invoke(Object[] arguments)
2018-09-05T21:25:32.2565701Z    at System.Dynamic.UpdateDelegates.UpdateAndExecute1[T0,TRet](CallSite site, T0 arg0)
2018-09-05T21:25:32.2721702Z    at System.Management.Automation.Interpreter.DynamicInstruction`2.Run(InterpretedFrame frame)
2018-09-05T21:25:32.2721702Z    at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)
2018-09-05T21:25:32.2721702Z    --- End of inner exception stack trace ---
2018-09-05T21:25:32.2721702Z    at System.Management.Automation.Runspaces.PipelineBase.Invoke(IEnumerable input)
2018-09-05T21:25:32.2721702Z    at System.Management.Automation.PowerShell.Worker.ConstructPipelineAndDoWork(Runspace rs, Boolean performSyncInvoke)
2018-09-05T21:25:32.2721702Z    at System.Management.Automation.PowerShell.Worker.CreateRunspaceIfNeededAndDoWork(Runspace rsToUse, Boolean isSync)
2018-09-05T21:25:32.2721702Z    at System.Management.Automation.PowerShell.CoreInvokeHelper[TInput,TOutput](PSDataCollection`1 input, PSDataCollection`1 output, PSInvocationSettings settings)
2018-09-05T21:25:32.2721702Z    at System.Management.Automation.PowerShell.CoreInvoke[TInput,TOutput](PSDataCollection`1 input, PSDataCollection`1 output, PSInvocationSettings settings)
2018-09-05T21:25:32.2721702Z    at System.Management.Automation.PowerShell.Invoke(IEnumerable input, PSInvocationSettings settings)
2018-09-05T21:25:32.2721702Z    at Microsoft.VisualStudio.Services.DevTestLabs.RemoteDeployer.Deployment.PowershellExecutor.<>c__DisplayClass28_0.<Invoke>b__0()
2018-09-05T21:25:32.2721702Z    at Microsoft.VisualStudio.Services.DevTestLabs.RemoteDeployer.Utilities.RetryExecutor.<>c__DisplayClass19_0`1.<Execute>b__0()
2018-09-05T21:25:32.2721702Z    at Microsoft.VisualStudio.Services.DevTestLabs.RemoteDeployer.Utilities.RetryExecutor.Execute(Action action)
2018-09-05T21:25:32.2721702Z    at Microsoft.VisualStudio.Services.DevTestLabs.RemoteDeployer.Utilities.RetryExecutor.Execute[T](Func`1 action)
2018-09-05T21:25:32.2721702Z    at Microsoft.VisualStudio.Services.DevTestLabs.RemoteDeployer.Deployment.PowershellExecutor.Invoke(String errorContextMessage, Boolean writeResultToLog, Boolean isCancellable, ActionPreference errorActionPreference)<---
```

### Log from Third Deployment

```
2018-09-06T15:35:22.1644549Z ==============================================================================
2018-09-06T15:35:22.1644549Z Task         : PowerShell on Target Machines
2018-09-06T15:35:22.1644549Z Description  : Execute PowerShell scripts on remote machine(s)
2018-09-06T15:35:22.1644549Z Version      : 2.0.3
2018-09-06T15:35:22.1644549Z Author       : Microsoft Corporation
2018-09-06T15:35:22.1644549Z Help         : [More Information](https://go.microsoft.com/fwlink/?linkid=627414)
2018-09-06T15:35:22.1800550Z ==============================================================================
2018-09-06T15:35:24.2860685Z Deployment started for machine: '<<<TARGET AZURE MACHINE OMITTED>>>:5986'
2018-09-06T15:35:30.1361060Z Deployment logs for Deployment operation on <<<TARGET AZURE MACHINE OMITTED>>>:5986 
2018-09-06T15:35:30.1361060Z 
2018-09-06T15:35:30.1361060Z 	 Deployment started on target machine...
2018-09-06T15:35:30.1361060Z 
2018-09-06T15:35:30.1361060Z System.AggregateException: Failed to execute the powershell script. Consult the logs below for details of the error.
2018-09-06T15:35:30.1361060Z The service has not been started. (Exception from HRESULT: 0x80070426) ---> System.Management.Automation.RuntimeException: The service has not been started. (Exception from HRESULT: 0x80070426) ---> System.Runtime.InteropServices.COMException: The service has not been started. (Exception from HRESULT: 0x80070426)
2018-09-06T15:35:30.1361060Z    at Microsoft.IIs.PowerShell.Framework.Interop.IAppHostMethodInstance.Execute()
2018-09-06T15:35:30.1361060Z    at Microsoft.IIs.PowerShell.Framework.CodeMethod.Invoke(Object[] arguments)
2018-09-06T15:35:30.1361060Z    at System.Dynamic.UpdateDelegates.UpdateAndExecute1[T0,TRet](CallSite site, T0 arg0)
2018-09-06T15:35:30.1361060Z    at System.Management.Automation.Interpreter.DynamicInstruction`2.Run(InterpretedFrame frame)
2018-09-06T15:35:30.1361060Z    at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)
2018-09-06T15:35:30.1361060Z    --- End of inner exception stack trace ---
2018-09-06T15:35:30.1361060Z    at System.Management.Automation.Runspaces.PipelineBase.Invoke(IEnumerable input)
2018-09-06T15:35:30.1361060Z    at System.Management.Automation.PowerShell.Worker.ConstructPipelineAndDoWork(Runspace rs, Boolean performSyncInvoke)
2018-09-06T15:35:30.1361060Z    at System.Management.Automation.PowerShell.Worker.CreateRunspaceIfNeededAndDoWork(Runspace rsToUse, Boolean isSync)
2018-09-06T15:35:30.1361060Z    at System.Management.Automation.PowerShell.CoreInvokeHelper[TInput,TOutput](PSDataCollection`1 input, PSDataCollection`1 output, PSInvocationSettings settings)
2018-09-06T15:35:30.1361060Z    at System.Management.Automation.PowerShell.CoreInvoke[TInput,TOutput](PSDataCollection`1 input, PSDataCollection`1 output, PSInvocationSettings settings)
2018-09-06T15:35:30.1361060Z    at System.Management.Automation.PowerShell.Invoke(IEnumerable input, PSInvocationSettings settings)
2018-09-06T15:35:30.1361060Z    at Microsoft.VisualStudio.Services.DevTestLabs.RemoteDeployer.Deployment.PowershellExecutor.<>c__DisplayClass28_0.<Invoke>b__0()
2018-09-06T15:35:30.1361060Z    at Microsoft.VisualStudio.Services.DevTestLabs.RemoteDeployer.Utilities.RetryExecutor.<>c__DisplayClass19_0`1.<Execute>b__0()
2018-09-06T15:35:30.1361060Z    at Microsoft.VisualStudio.Services.DevTestLabs.RemoteDeployer.Utilities.RetryExecutor.Execute(Action action)
2018-09-06T15:35:30.1361060Z    at Microsoft.VisualStudio.Services.DevTestLabs.RemoteDeployer.Utilities.RetryExecutor.Execute[T](Func`1 action)
2018-09-06T15:35:30.1361060Z    at Microsoft.VisualStudio.Services.DevTestLabs.RemoteDeployer.Deployment.PowershellExecutor.Invoke(String errorContextMessage, Boolean writeResultToLog, Boolean isCancellable, ActionPreference errorActionPreference)
2018-09-06T15:35:30.1517061Z    --- End of inner exception stack trace ---
2018-09-06T15:35:30.1517061Z    at Microsoft.VisualStudio.Services.DevTestLabs.RemoteDeployer.Deployment.PowershellExecutor.Invoke(String errorContextMessage, Boolean writeResultToLog, Boolean isCancellable, ActionPreference errorActionPreference)
2018-09-06T15:35:30.1517061Z    at Microsoft.VisualStudio.Services.DevTestLabs.RemoteDeployer.Deployment.ScriptExecutor.ExecuteScript(IPowerShell powerShell, String script, String arguments, Boolean isScriptDotSourced, Boolean useHttp)
2018-09-06T15:35:30.1517061Z    at Microsoft.VisualStudio.Services.DevTestLabs.RemoteDeployer.Deployment.DeploymentService.RunPowerShellScripts(DeploymentMachineSpecification deploymentMachine, ScriptSpecification scriptSpecification, ScriptSpecification initializationScriptSpecification, String applicationPath, IPowerShell powerShellSession)
2018-09-06T15:35:30.1517061Z ---> (Inner Exception #0) System.Management.Automation.RuntimeException: The service has not been started. (Exception from HRESULT: 0x80070426) ---> System.Runtime.InteropServices.COMException: The service has not been started. (Exception from HRESULT: 0x80070426)
2018-09-06T15:35:30.1517061Z    at Microsoft.IIs.PowerShell.Framework.Interop.IAppHostMethodInstance.Execute()
2018-09-06T15:35:30.1517061Z    at Microsoft.IIs.PowerShell.Framework.CodeMethod.Invoke(Object[] arguments)
2018-09-06T15:35:30.1517061Z    at System.Dynamic.UpdateDelegates.UpdateAndExecute1[T0,TRet](CallSite site, T0 arg0)
2018-09-06T15:35:30.1517061Z    at System.Management.Automation.Interpreter.DynamicInstruction`2.Run(InterpretedFrame frame)
2018-09-06T15:35:30.1517061Z    at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)
2018-09-06T15:35:30.1517061Z    --- End of inner exception stack trace ---
2018-09-06T15:35:30.1517061Z    at System.Management.Automation.Runspaces.PipelineBase.Invoke(IEnumerable input)
2018-09-06T15:35:30.1517061Z    at System.Management.Automation.PowerShell.Worker.ConstructPipelineAndDoWork(Runspace rs, Boolean performSyncInvoke)
2018-09-06T15:35:30.1517061Z    at System.Management.Automation.PowerShell.Worker.CreateRunspaceIfNeededAndDoWork(Runspace rsToUse, Boolean isSync)
2018-09-06T15:35:30.1517061Z    at System.Management.Automation.PowerShell.CoreInvokeHelper[TInput,TOutput](PSDataCollection`1 input, PSDataCollection`1 output, PSInvocationSettings settings)
2018-09-06T15:35:30.1517061Z    at System.Management.Automation.PowerShell.CoreInvoke[TInput,TOutput](PSDataCollection`1 input, PSDataCollection`1 output, PSInvocationSettings settings)
2018-09-06T15:35:30.1517061Z    at System.Management.Automation.PowerShell.Invoke(IEnumerable input, PSInvocationSettings settings)
2018-09-06T15:35:30.1517061Z    at Microsoft.VisualStudio.Services.DevTestLabs.RemoteDeployer.Deployment.PowershellExecutor.<>c__DisplayClass28_0.<Invoke>b__0()
2018-09-06T15:35:30.1517061Z    at Microsoft.VisualStudio.Services.DevTestLabs.RemoteDeployer.Utilities.RetryExecutor.<>c__DisplayClass19_0`1.<Execute>b__0()
2018-09-06T15:35:30.1517061Z    at Microsoft.VisualStudio.Services.DevTestLabs.RemoteDeployer.Utilities.RetryExecutor.Execute(Action action)
2018-09-06T15:35:30.1517061Z    at Microsoft.VisualStudio.Services.DevTestLabs.RemoteDeployer.Utilities.RetryExecutor.Execute[T](Func`1 action)
2018-09-06T15:35:30.1517061Z    at Microsoft.VisualStudio.Services.DevTestLabs.RemoteDeployer.Deployment.PowershellExecutor.Invoke(String errorContextMessage, Boolean writeResultToLog, Boolean isCancellable, ActionPreference errorActionPreference)<---
```

**Comment by RoshanKumarMicrosoft** [comment]
@qJake  The error message could result due to the following reason:
1. The service associated with Credential Manager does not start.
2.  Some files associated with the application have gone corrupt.
Can you try your custom script locally.? It looks like there is some issue in your custom script . If you want you can use our IIS deployment or IIS management task instead of writing your custom script .

**Comment by qJake** [comment]
Hi @RoshanKumarMicrosoft,

First, I didn't want to use the IIS Deployment task as it is deprecated.

Second, my PowerShell script is *extremely* simple, it does not have any connectivity by itself, it just moves files and does things locally:

```powershell
param
(
    [string] $AppPool,
    [string] $Source,
    [string] $Dest
)

if (!$AppPool)
{
    throw 'AppPool parameter is missing!';
}

if (!$Source)
{
    throw 'Source parameter is missing!';
}

if (!$Dest)
{
    throw 'Dest parameter is missing!';
}

if(!(Test-Path $Source))
{
    throw "Source path $Source does not exist!"
}

if(!(Test-Path $Dest))
{
    throw "Destination path $Dest does not exist!"
}

# Imports
Import-Module WebAdministration

# Stop app pool
$pool = (gi "IIS:\AppPools\$AppPool")
$pool.Stop()

# Wait a moment for it to stop
sleep 2

# Clean destination
rm "$Dest\*" -Recurse -Force

# Copy files
cp $Source $Dest -Recurse -Force

# Start app pool
$pool.Start()
```

What else could it be? If we suspect that it's something specific to Azure VMs, I would need something concrete to bring to the Azure team or support team so that they can help debug it.

**Comment by RoshanKumarMicrosoft** [comment]
@qJake  This is [IIS Web app Deployment through winrm](https://marketplace.visualstudio.com/items?itemName=ms-vscs-rm.iiswebapp) which is available as extension. You can also use [IIS web app deployment task for deployment group](https://github.com/Microsoft/vsts-tasks/tree/master/Tasks/IISWebAppDeploymentOnMachineGroupV0). 

Can you also please try with power shell on target machine having version 3 ? Steps to upload  powershell on target machine task version 3 on tfs .
You can either use cmd or powershell to run the commands.
1. Clone the vsts tasks repo ( git clone https://github.com/Microsoft/vsts-tasks ).
2. cd into this repo.
3. run: npm install
4. run: npm install -g tfx-cli
5. run: gulp build --task PowerShellOnTargetMachinesV3
6. run: tfx login
   ( The above command will prompt you for the url of your TFS instance. The Url should contain the collection name as well. For e.g. if the machine where
     the TFS is installed is 'foobar' and your default collection is named 'DefaultCollection', then the url provided should be: 
     'http://foobar:8080/tfs/DefaultCollection' . The command will also prompt you for a PAT, which you will need to generate from the security tab in TFS)
7. run: tfx build tasks upload --task-path .\_build\Tasks\PowerShellOnTargetMachinesV3
8. The task should now be uploaded and you should see it as part of the task dropdown.
9. run: tfx logout

If you still face any issue please share your SkypeId on RM_Customer_Queries@microsoft.com.

**Comment by chshrikh** [comment]
@qJake are you unblocked on this

**Comment by qJake** [comment]
Hi @chshrikh I have not had a chance to test the actual V3 Preview task but it did install successfully.

If you would like, you can close this issue, and I can reopen this one or open a new issue if the V3 step fails with the same error after I have had a chance to test it.
